pkg/graveler/ref/commit_walker.go

```diff
@@ -68,34 +68,38 @@ func (w *CommitWalker) Err() error {
 	return w.err
 }
 
-func FindLowestCommonAncestor(ctx context.Context, getter CommitGetter, addressProvider ident.AddressProvider, repositoryID graveler.RepositoryID, left, right graveler.CommitID) (*graveler.Commit, error) {
-	discoveredSet := make(map[string]struct{})
+func FindLowestCommonAncestor(ctx context.Context, getter CommitGetter, addressProvider ident.AddressProvider, repositoryID graveler.RepositoryID, left, right graveler.CommitID) *graveler.Commit {
 	iterLeft := NewCommitWalker(ctx, getter, repositoryID, left)
+	// find the set of all ancestors of the left commit:
+	leftAncestors := make(map[string]bool)
+	for iterLeft.Next() {
+		addr := addressProvider.ContentAddress(iterLeft.Value())
+		leftAncestors[addr] = true
+	}
 	iterRight := NewCommitWalker(ctx, getter, repositoryID, right)
-	for {
-		commit, err := findLowestCommonAncestorNextIter(addressProvider, discoveredSet, iterLeft)
-		if commit != nil || err != nil {
-			return commit, err
-		}
-		commit, err = findLowestCommonAncestorNextIter(addressProvider, discoveredSet, iterRight)
-		if commit != nil || err != nil {
-			return commit, err
+	// find the set of all common ancestors:
+	commonAncestors := make(map[string]*graveler.Commit)
+	for iterRight.Next() {
+		commit := iterRight.Value()
+		addr := addressProvider.ContentAddress(commit)
+		if leftAncestors[addr] {
+			commonAncestors[addr] = commit
 		}
-		if iterLeft.Value() == nil && iterRight.Value() == nil {
-			break
+	}
```

Comment:
Just to make sure I understand the algorithm:

1. scan all left ancestors
2. scan all right ancestors, comparing to the left set to derive the common ones
3. for each common ancesstor, increment a counter for its parents
4. find and return the first arbitrarily found common ancestor that has no child node that is also a common ancestor

so time: O(N+M) where N is left ancestors and M is right ancestors
space: O(N) because we keep a set of all leftAncestors

My worries:
1. whether this will always result in a correct lowest common ancestor
2. the high cost involved with the solution.

Not claiming I have a simpler one in mind, I think it is a rather complex problem to solve efficiently - but I feel this is a problem that has already received quite a lot of attention as it is solved in Git. I would recommend either adopting their solution, or sticking with this one - with a clear explanation of why it's better for lakeFS.

Reply:
You understand correctly. 
This problem didn't get as much attention as you would expect. For trees yes, but not for DAGs.

Your worry (1) can be solved using documentation.

This is absolutely not documented in git, but if I understand the code correctly, they solve this using a priority-queue, where the priority function is the generation of the commit (generation is the distance from the root). I will look into it.

Reply:
A good post I found: https://devblogs.microsoft.com/devops/supercharging-the-git-commit-graph-iii-generations/